Source/WebCore/html/HTMLAttachmentElement.cpp

```diff
@@ -146,28 +175,72 @@ void HTMLAttachmentElement::ensureModernShadowTree(ShadowRoot& root)
     m_innerLegacyAttachment->m_thumbnail = WTFMove(m_thumbnail);
     m_innerLegacyAttachment->m_icon = WTFMove(m_icon);
     m_innerLegacyAttachment->m_iconSize = m_iconSize;
-    m_innerLegacyAttachment->setIdAttribute(attachmentPreviewIdentifier());
-    container->appendChild(*m_innerLegacyAttachment);
+    addShadowElement(m_innerLegacyAttachment, attachmentPreviewIdentifier(), String { });
+
+    m_elementWithAction = createShadowDivElement(attachmentActionIdentifier(), attachmentActionForDisplay());
+
+    m_elementWithTitle = createShadowDivElement(attachmentTitleIdentifier(), attachmentTitleForDisplay());
+
+    m_elementWithSubtitle = createShadowDivElement(attachmentSubtitleIdentifier(), attachmentSubtitleForDisplay());
+
+    updateSave(attributeWithoutSynchronization(saveQualifiedName()));
+}
 
-    m_elementWithAction = HTMLDivElement::create(document());
-    m_elementWithAction->setIdAttribute(attachmentActionIdentifier());
-    if (const auto& action = attachmentActionForDisplay(); !action.isEmpty())
-        m_elementWithAction->setInnerText(String { action });
-    container->appendChild(*m_elementWithAction);
+class AttachmentSaveEventListener final : public EventListener {
+public:
+    static Ref<AttachmentSaveEventListener> create(HTMLAttachmentElement& attachment) { return adoptRef(*new AttachmentSaveEventListener(attachment)); }
 
-    m_elementWithTitle = HTMLDivElement::create(document());
-    m_elementWithTitle->setIdAttribute(attachmentTitleIdentifier());
-    if (auto title = attachmentTitleForDisplay(); !title.isEmpty())
-        m_elementWithTitle->setInnerText(WTFMove(title));
-    container->appendChild(*m_elementWithTitle);
+    bool operator==(const EventListener& other) const final
+    {
+        return this == &other;
+    }
 
-    m_elementWithSubtitle = HTMLDivElement::create(document());
-    m_elementWithSubtitle->setIdAttribute(attachmentSubtitleIdentifier());
-    if (auto subtitle = attachmentSubtitleForDisplay(); !subtitle.isEmpty())
-        m_elementWithSubtitle->setInnerText(WTFMove(subtitle));
-    container->appendChild(*m_elementWithSubtitle);
+    void handleEvent(ScriptExecutionContext&, Event& event) final
+    {
+        if (event.type() == eventNames().clickEvent) {
+            event.preventDefault();
+            event.stopPropagation();
+            event.stopImmediatePropagation();
```

Comment:
I assume this is to prevent the default attachment click behavior, but are all three of these needed?

Reply:
I don't know, I blindly copied it from another event listener (e.g. [`MediaControlsContextMenuEventListener::handleEvent`](https://searchfox.org/wubkat/rev/dacbd7c36056647c1867f34236ade6815002e1f1/Source/WebCore/Modules/mediacontrols/MediaControlsHost.cpp#455-457)). 🤷

Reply:
Usually these are trying to prevent the default click event listener from triggering, so they should only be needed if the attachment element itself has a click listener, and you don't want it to run it at the same time as the save button listener.

Reply:
That's the case here: If the user clicks on the save button, it gets converted into a custom event out of the top-level attachment, and I don't want the initial click to invoke click event listeners on that attachment.